### DISCUSSION TITLE
Transformations of checks - typescript, good parts

### DISCUSSION THREAD

**Discussion by happy-barney** [Perl-Oshun/oshun#17] [Ideas]
For later milestones, created just to not forget.

[p5p post](https://www.nntp.perl.org/group/perl.perl5.porters/2023/05/msg266478.html)

Inspired by typescript:
- [Utility types](https://www.typescriptlang.org/docs/handbook/utility-types.html)
- [Type transformation](https://www.typescriptlang.org/docs/handbook/2/types-from-types.html)

It will be nice to have some kind of sub-language allowing to transform existing checks into complementary checks. For example, validation of input data:

Example (pseudocode):

```perl
declare Item => Dict [ login => Str, email => Email ];
my @request_content :of ([ keyof Item ] => Optional[ Str ]) = $request->content;

for my $item (@request_content) {
  if (validate ($request :is (Item)) {
    push @response, process_item ($item);
  } else {
    push @response, validation_error (...);
  }
}
```

**Comment by zmughal** [comment]
Some more utilities that are popular: <https://github.com/millsp/ts-toolbelt>.

**Comment by jjn1056** [comment]
I don't understand 

validate ($request :is (Item)

when calling a method or function usually in strongly typed languages you just pass the variable.  the only time you might add something is if you are casting or coercing.

**Reply by happy-barney** [reply]
but perl isn't and will not be strongly typed ... Oshun only may offer you to do it but you have to respect another programmers, another approaches.

in this case, it describes handling of external data - you don't want send customer your internal error but you want to explain it, format it.

so you can read it as:
- we have "strongly" typed `Item`
- customer sends something
- we validate that and politely points out what is wrong

**Comment by druud** [comment]
Perl is "strongly typed", but in different ways than with most other languages. There are strong type attributes in sigils, and in operators. There are also plenty of historical errors in acknowledging that, which then always need to be reverted or otherwise fixed.
